alternator/executor.cc

```diff
@@ -2358,21 +2358,22 @@ std::optional<rjson::value> executor::describe_single_item(schema_ptr schema,
     return item;
 }
 
-std::vector<rjson::value> executor::describe_multi_item(schema_ptr schema,
-        const query::partition_slice& slice,
-        const cql3::selection::selection& selection,
-        const query::result& query_result,
-        const std::optional<attrs_to_get>& attrs_to_get) {
-    cql3::selection::result_set_builder builder(selection, gc_clock::now(), cql_serialization_format::latest());
-    query::result_view::consume(query_result, slice, cql3::selection::result_set_builder::visitor(builder, *schema, selection));
+future<std::vector<rjson::value>> executor::describe_multi_item(schema_ptr schema,
+        const query::partition_slice&& slice,
+        shared_ptr<cql3::selection::selection> selection,
+        foreign_ptr<lw_shared_ptr<query::result>> query_result,
+        shared_ptr<const std::optional<attrs_to_get>> attrs_to_get) {
+    cql3::selection::result_set_builder builder(*selection, gc_clock::now(), cql_serialization_format::latest());
+    query::result_view::consume(*query_result, slice, cql3::selection::result_set_builder::visitor(builder, *schema, *selection));
```

Comment:
While the result building non-preemptible time is now O(row), this consume() is still O(entire result).

Reply:
That's true, how is this solved (or not) for cql ? I ran `pytest -vv -s test_paging.py::test_sticky_stop_flag` and see that the whole `SELECT` batch runs in a single loop inside `result_view::consume`.

I'm not sure how to easily change those cql visitors (mentioned also a similar case in https://github.com/scylladb/scylladb/pull/12351#issuecomment-1357388012). To my current understanding I'd need to wrap every function result (and all ancestor callers) in future<...>. Also taking extra care that there is no lambda involved.

It seems easy to do in alternator but I'd like to avoid touching cql for now as my code understanding is very limited there. Or am I missing some simpler way?

Reply:
It's probably not solved. I don't even know if we need to solve it. For sure maybe-yielding everywhere will regress the common case when just a few rows are returned, so we'll have to special-case them.